public/js/util/sources-tree.js

```diff
@@ -59,59 +62,167 @@ function getURL(source) {
   return url;
 }
 
-function addToTree(tree, source) {
+/**
+ * Takes two parts array, and returns the common path.
+ */
+function getSharedParts(partsA, partsB) {
+  const sharedParts = [];
+  for (let j = 0; j < partsA.length; j++) {
+    const partA = partsA[j];
+    const partB = partsB[j];
+    if (partA === partB) {
+      sharedParts.push(partA);
+    } else {
+      break;
+    }
+  }
+  return sharedParts;
+}
+
+function addToTree(currentNode, source) {
   const url = getURL(source);
   if (!url) {
     return;
   }
 
   const parts = url.pathname.split("/").filter(p => p !== "");
-  const isDir = (parts.length === 0 ||
-                 parts[parts.length - 1].indexOf(".") === -1);
+  const isDir = partsIsDir(parts);
   parts.unshift(url.host);
 
-  let path = "";
-  let subtree = tree;
+  let accumulatedPath = "";
+
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    const contents = currentNode.contents;
 
-  for (let part of parts) {
     // Currently we assume that we are descending into a node with
     // children. This will fail if a path has a directory named the
     // same as another file, like `foo/bar.js/file.js`.
     //
-    // TODO: Be smarter about this, which we'll probably do when we
-    // are smarter about folders and collapsing empty ones.
-    assert(nodeHasChildren(subtree), `${subtree.name} should have children`);
-    const subpaths = subtree.contents;
+    // TODO: Be smarter about this.
+    assert(nodeHasChildren(currentNode),
+           `${currentNode.name} should have children`);
 
     // We want to sort alphabetically, so find the index where we
     // should insert this part.
-    let idx = subpaths.findIndex(subpath => {
-      return subpath.name.localeCompare(part) >= 0;
-    });
-
-    if (idx >= 0 && subpaths[idx].name === part) {
-      // A node with the same name already exists, simply traverse
-      // into it.
-      subtree = subpaths[idx];
+    let index = getSortedIndex(contents, part);
+
+    if (index >= 0 && contents[index].baseName === part) {
+      // A node with the same name already exists
+      if (contents[index].name !== contents[index].baseName) {
```

Comment:
This code is a bit complex (inherently so), so I'm looking for any way to make it easier to follow. What do you think about making a `isCollapsed` function that runs his logic? You use it below as well. I don't care about abstracting out this logic as much as making this workflow easier to follow.